sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala

```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
+      assert(client.getPartitionOption("default", "src_part", spec).isEmpty)
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Function related API
+    ///////////////////////////////////////////////////////////////////////////
+
+    def function(name: String, className: String): CatalogFunction = {
+      CatalogFunction(
+        FunctionIdentifier(name, Some("default")), className, Seq.empty[FunctionResource])
+    }
+
+    test(s"$version: createFunction") {
+      val functionClass = "org.apache.spark.MyFunc1"
+      client.createFunction("default", function("func1", functionClass))
+    }
+
+    test(s"$version: functionExists") {
+      assert(client.functionExists("default", "func1") == true)
+    }
+
+    test(s"$version: renameFunction") {
+      client.renameFunction("default", "func1", "func2")
+      assert(client.functionExists("default", "func2") == true)
+    }
+
+    test(s"$version: alterFunction") {
+      val functionClass = "org.apache.spark.MyFunc2"
+      client.alterFunction("default", function("func2", functionClass))
+    }
+
+    test(s"$version: getFunction") {
+      // No exception should be thrown
+      val func = client.getFunction("default", "func2")
+      assert(func.className == "org.apache.spark.MyFunc2")
+    }
+
+    test(s"$version: getFunctionOption") {
+      assert(client.getFunctionOption("default", "func2").isDefined)
+    }
+
+    test(s"$version: listFunctions") {
+      assert(client.listFunctions("default", "fun.*").size == 1)
+    }
+
+    test(s"$version: dropFunction") {
+      client.dropFunction("default", "func2")
+      assert(client.listFunctions("default", "fun.*").size == 0)
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // SQL related API
+    ///////////////////////////////////////////////////////////////////////////
+
+    test(s"$version: sql set command") {
+      client.runSqlHive("SET spark.sql.test.key=1")
+    }
+
+    test(s"$version: sql create index and reset") {
       client.runSqlHive("CREATE TABLE indexed_table (key INT)")
       client.runSqlHive("CREATE INDEX index_1 ON TABLE indexed_table(key) " +
         "as 'COMPACT' WITH DEFERRED REBUILD")
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Miscellaneous API
+    ///////////////////////////////////////////////////////////////////////////
+
+    test(s"$version: version") {
+      client.version.fullVersion
+    }
+
+    test(s"$version: getConf") {
+      assert("success" === client.getConf("test", null))
+    }
+
+    test(s"$version: setOut") {
+      client.setOut(new PrintStream(new ByteArrayOutputStream()))
+    }
+
+    test(s"$version: setInfo") {
+      client.setInfo(new PrintStream(new ByteArrayOutputStream()))
+    }
+
+    test(s"$version: setError") {
+      client.setError(new PrintStream(new ByteArrayOutputStream()))
+    }
```

Comment:
For `setOut`/`setInfo`/`setError`, it is fine to not have tests. They are not actually calling any method of hive metastore client.